### PR TITLE
Release v1.0.0 beta.4

### PR DESCRIPTION
**Pull request by jvff**
---
name: Release Checklist Template
about: Checklist of versioning to create a taggable commit for Zebra
title: ''
labels:
assignees: ''

---

### Which Crates to Increment

To check if any of the top-level crates need version increments, go to the zebra GitHub code page: https://github.com/ZcashFoundation/zebra and use the last modified dates of each crate. Alternatively you can use the github compare tool and check the `main` branch against the last tag ([Example](https://github.com/ZcashFoundation/zebra/compare/v1.0.0-alpha.15...main)). `git diff --stat <previous-release-tag> origin/main` is also useful to see what's changed.

- [x] Increment the crates that have new commits since the last version update
- [x] Increment any crates that depend on crates that have changed
- [x] Use the `zebrad` crate version in the `zebra-network` user agent string
      (currently the constant `USER_AGENT` in `zebra-network/src/constants.rs`)
- [x] Use the latest git tag in `README.md`
- [x] Use the latest git tag in `book/src/user/install.md`

### How to Increment Versions

Zebra follows [semantic versioning](https://semver.org).

Semantic versions look like: `MAJOR`.`MINOR`.`PATCH[`-`TAG`.`PRE-RELEASE]`

#### Pre-Release Crates

Pre-Release versions have a `TAG` like "alpha" or "beta". For example: `1.0.0-alpha.0`

1. Increment the `PRE-RELEASE` version for the crate.

Optionally, if a `MINOR` feature pre-release breaks `MAJOR` API compatibility:

2. Increment the `MAJOR` version, and reset all the other versions to zero

#### Unstable Crates

Unstable versions have a `MAJOR` version of zero. For example: `0.1.0`

1. Follow stable crate versioning, but increment the `MINOR` version for breaking changes

#### Stable Crates

For example: `1.0.0`

Increment the first version component in this list, and reset the other components to zero:
1. MAJOR versions for breaking public API changes and removals
    * check for types from dependencies that appear in the public API
2. MINOR versions for new features
3. PATCH versions for bug fixes
    * includes dependency updates that don't impact the public API

### Version Locations

Once you know which versions you want to increment, you can find them in the:
- [x] zebra* `Cargo.toml`s
- [x] tower-* `Cargo.toml`s
- [x] `zebra-network` protocol user agent: https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/constants.rs
- [x] `README.md`
- [x] `book/src/user/install.md`
- [x] `Cargo.lock`: automatically generated by `cargo build`

#### Version Tooling

You can use `fastmod` to interactively find and replace versions.

For example, you can do something like:
```
fastmod --extensions rs,toml,md --fixed-strings '1.0.0-alpha.12' '1.0.0-alpha.13'
fastmod --extensions rs,toml,md --fixed-strings '0.2.9' '0.2.10' tower-batch
fastmod --extensions rs,toml,md --fixed-strings '0.2.8' '0.2.9' tower-fallback
```

### Reviewing Version Bumps

Check for missed changes by going to:
`https://github.com/ZcashFoundation/zebra/tree/<commit-hash>/`
Where `<commit-hash>` is the hash of the last commit in the version bump PR.

If any Zebra or Tower crates have commit messages that are **not** a version bump, we have missed an update.

Also check for crates that depend on crates that have changed. They should get a version bump as well.

## Initial Testing

- [x] After any changes, test that the `cargo install` command in works. Use
      e.g. `cargo install --locked --path zebrad`.

## README

As we resolve various outstanding known issues and implement new functionality with each release, we should double check the README for any necessary updates.

We should check and update if necessary:

- [x] The "Beta Release" section
- [x] The "Known Issues" section

to ensure that any items that are resolved in the latest release are no longer listed in the README.

## Change Log

**Important**: Any merge into `main` deletes any edits to the draft changelog. Once you are ready to tag a release, copy the draft changelog into `CHANGELOG.md`.

We follow the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

We use [the Release Drafter workflow](https://github.com/marketplace/actions/release-drafter) to automatically create a [draft changelog](https://github.com/ZcashFoundation/zebra/releases).

To create the final change log:
- [x] Copy the draft changelog into `CHANGELOG.md`
- [x] Delete any trivial changes
- [x] Combine duplicate changes
- [x] Edit change descriptions so they are consistent, and make sense to non-developers
- [x] Check the category for each change
  - prefer the "Fix" category if you're not sure

#### Change Categories

From "Keep a Changelog":
* `Added` for new features.
* `Changed` for changes in existing functionality.
* `Deprecated` for soon-to-be removed features.
* `Removed` for now removed features.
* `Fixed` for any bug fixes.
* `Security` in case of vulnerabilities.

## Create the release PR

After you have the version increments and the updated changelog,

- [x] Push the version increments and the updated changelog into a branch
      (name suggestion, example: `v1.0.0-alpha.0-release`)
- [x] Create a release PR by adding `&template=release-checklist.md` to the
      comparing url ([Example](https://github.com/ZcashFoundation/zebra/compare/v1.0.0-alpha.0-release?expand=1&template=release-checklist.md)).
- [x] While the PR is being reviewed, keep track of any PRs that have been merged
      since you created the PR to update `CHANGELOG.md` and push any updates if necessary
- [x] Once the release PR has been approved and merged, create a new release
      using the draft release as a base, by clicking the Edit icon
      in the [draft release](https://github.com/ZcashFoundation/zebra/releases).
- [x] Set the release title to `Zebra ` followed by the version tag, for example:
      `Zebra 1.0.0-alpha.0`
- [x] Copy the final changelog of this release to the release description
      (starting just _after_ the title `## [Zebra ...`)
- [x] Set the tag name to the version tag, for example: `v1.0.0-alpha.0`
- [x] Set the release to target the `main` branch
- [x] Mark the release as 'pre-release' (until we are no longer alpha/beta)
- [x] Publish the release

## Final Testing

- [ ] After tagging the release, test that the exact `cargo install` command in
      `README.md` works (`--git` behaves a bit differently to `--path`)

If the build fails after tagging:
1. fix the build
2. check if the fixes changed any extra crates, and do the required version increments
3. update `README.md` with a **new** git tag
4. update `CHANGELOG.md` with details about the fix
5. tag a **new** release

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,64 @@ All notable changes to Zebra are documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org).
 
+## [Zebra 1.0.0-beta.4](https://github.com/ZcashFoundation/zebra/releases/tag/v1.0.0-beta.4) - 2022-01-26
+
+Zebra's latest beta improves the networking code and fixes some bugs. A couple of fixed bugs had
+caused Zebra to hang in some situations. Some improvements to the documentation were also included.
```
Comment (teor2345) [inline CHANGELOG.md:10]: Do we want to mention Rust 2021 edition here?

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,64 @@ All notable changes to Zebra are documented in this file.
...
+- Update to use Rust edition 2021 (#3332)
+
+#### Networking
+
+- Cache incoming unsolicited address messages, and use them as responses (#3294)
+- Cleanup internal network request handler, fix unused request logging (#3295)
+
+#### Documentation
+
+- Document the consensus rules for Section 3.6 (#3338)
```
Comment (teor2345) [inline CHANGELOG.md]: @upbqdn can you add the section title to this changelog entry?

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,64 @@ All notable changes to Zebra are documented in this file.
...
+- Make the Zebra state RFC match the lightwalletd draft design (#3385)
```
Comment (teor2345) [inline CHANGELOG.md]: This is a revert
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -105,7 +105,7 @@ for your platform:
 2. Install Zebra's build dependencies:
      - **libclang:** the `libclang`, `libclang-dev`, `llvm`, or `llvm-dev` packages, depending on your package manager
      - **clang** or another C++ compiler: `g++`, `Xcode`, or `MSVC`
-3. Run `cargo install --locked --git https://github.com/ZcashFoundation/zebra --tag v1.0.0-beta.3 zebrad`
+3. Run `cargo install --locked --git https://github.com/ZcashFoundation/zebra --tag 1.0.0-beta.4 zebrad`
```
Comment (teor2345) [inline README.md]: Dropping the initial `v` looks like a typo:
```suggestion
3. Run `cargo install --locked --git https://github.com/ZcashFoundation/zebra --tag v1.0.0-beta.4 zebrad`
```

File: book/src/user/install.md [hunk 5]
```diff
@@ -9,7 +9,7 @@ for your platform:
 2. Install Zebra's build dependencies:
      - **libclang:** the `libclang`, `libclang-dev`, `llvm`, or `llvm-dev` packages, depending on your package manager
      - **clang** or another C++ compiler: `g++`, `Xcode`, or `MSVC`
-3. Run `cargo install --locked --git https://github.com/ZcashFoundation/zebra --tag v1.0.0-beta.3 zebrad`
+3. Run `cargo install --locked --git https://github.com/ZcashFoundation/zebra --tag 1.0.0-beta.4 zebrad`
```
Comment (teor2345) [inline book/src/user/install.md]:
```suggestion
3. Run `cargo install --locked --git https://github.com/ZcashFoundation/zebra --tag v1.0.0-beta.4 zebrad`
```

**Review by teor2345 — COMMENTED** [review]
Looks good, but I'd like someone else to review it too.

File: CHANGELOG.md (hunk 1 above)
Comment (jvff) [reply CHANGELOG.md:10]: Done.

File: CHANGELOG.md (hunk 2 above)
Comment (jvff) [reply CHANGELOG.md]: Done, let me know if it's okay.

Comment (upbqdn) [reply CHANGELOG.md]: Looks good.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,64 @@ All notable changes to Zebra are documented in this file.
...
+#### Networking
+
+- Cache incoming unsolicited address messages, and use them as responses (#3294)
+- Cleanup internal network request handler, fix unused request logging (#3295)
+
+#### Documentation
+
+- Document the consensus rules for Spend Transfers, Output Transfers, and their Descriptions (Section 3.6 of the Zcash protocol specification) (#3338)
+- Document the structure of the zebra-network crate (#3317)
+- Document the consensus rules for Sec. 3.8 Note Commitment Trees (#3319)
```
Comment (upbqdn) [inline CHANGELOG.md]:
```suggestion
- Document the consensus rules for Note Commitment Trees (Section 3.8 of the Zcash Protocol Specification) (#3319)
```
This suggestion makes the entries related to the consensus rules consistent.

**Comment by upbqdn** [comment]
I think we can also remove

    Validation of Sprout anchors (root of the Sprout note commitment tree)

from `README.md`, and say we validate all consensus rules from the spec. 🎉

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,68 @@ All notable changes to Zebra are documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org).
 
+## [Zebra 1.0.0-beta.4](https://github.com/ZcashFoundation/zebra/releases/tag/v1.0.0-beta.4) - 2022-01-26
```
Comment (teor2345) [inline CHANGELOG.md:7]: Mergify will merge this PR as soon as there is an approval, so I'm leaving this comment to stop that happening until Janito is ready to tag.

Comment (jvff) [reply CHANGELOG.md:7]: I'm ready I believe. I mean, we can get it merged and I can tag it tomorrow, right?

**Review by teor2345 — APPROVED** [review]
Let's go, we can tag whenever

**Pull request merged**